### PR TITLE
Add about org

### PR DESCRIPTION
**Pull request by dankmitchell**
### Context
Enrich organisation information

### Changes proposed in this pull request
Add "About this organisation"

### Guidance to review
`/organisation/{ucaseCode}/about`

# After
screen shot 2018-08-01 at 14 11 26

### CODE REVIEW

**Comment by fofr** [comment]
@dankmitchell Could you add a screenshot?

**Comment by dankmitchell** [comment]
@fofr the page is so long I didn't both but will do it now

**Comment by fofr** [comment]
I notice the screenshot has the website field. We already have some website information from UCAS. Do these two connect up?

**Comment by timabell** [comment]
@fofr no.
`UcasInstitution.Url` is currently imported but not used anywhere.

**Comment by fofr** [comment]
Let's remove the website field for now, we'll use the UCAS data.

**Comment by timabell** [comment]
Reviewing...

File: src/ui/Startup.cs [hunk 1]
```diff
@@ -43,10 +46,67 @@ public void ConfigureServices(IServiceCollection services)
...
             {
-                options.ExpireTimeSpan = TimeSpan.FromMinutes(20);
+                options.ExpireTimeSpan = TimeSpan.FromMinutes(120);
+                options.Events = new CookieAuthenticationEvents
+                {
+                    // this event is fired everytime the cookie has been validated by the cookie middleware,
+                    // so basically during every authenticated request
+                    // the decryption of the cookie has already happened so we have access to the user claims
+                    // and cookie properties - expiration, etc..
+                    OnValidatePrincipal = async x =>
```
Comment (timabell) [inline src/ui/Startup.cs]: Quoting references would be handy for others trying to understand this later. Url in comment would be fine.

File: src/ui/Startup.cs [hunk 2]
```diff
@@ -43,10 +46,67 @@ public void ConfigureServices(IServiceCollection services)
...
+                        {
+                            var identity = (ClaimsIdentity)x.Principal.Identity;
+                            var accessTokenClaim = identity.FindFirst("access_token");
+                            var refreshTokenClaim = identity.FindFirst("refresh_token");
+
+                            // if we have to refresh, grab the refresh token from the claims, and request
+                            // new access token and refresh token
+                            var refreshToken = refreshTokenClaim.Value;
+
+                            var clientId = Configuration["auth:oidc:clientId"];
```
Comment (timabell) [inline src/ui/Startup.cs]: DRY

File: src/ui/Startup.cs [hunk 3]
```diff
@@ -43,10 +46,67 @@ public void ConfigureServices(IServiceCollection services)
...
+                            if (string.IsNullOrWhiteSpace(clientSecret))
+                            {
+                                throw new Exception("Missing environment variable " + envKeyClientSecret + " - get this from the DfE Sign-in team.");
+                            }
+                            var tokenEndpoint = Configuration["auth:oidc:tokenEndpoint"];
+
+                            var client = new TokenClient(tokenEndpoint, clientId, clientSecret);
+                            var response = await client.RequestRefreshTokenAsync(refreshToken, new {client_secret = clientSecret});
+
+                            if (!response.IsError)
```
Comment (timabell) [inline src/ui/Startup.cs]: invert for less nesting

File: src/ui/Startup.cs [hunk 4]
```diff
@@ -63,7 +123,11 @@ public void ConfigureServices(IServiceCollection services)
                 options.ClientSecret = clientSecret;
                 options.ResponseType = OpenIdConnectResponseType.Code;
                 options.GetClaimsFromUserInfoEndpoint = true;
-                options.UseTokenLifetime = true;
+
+                // using this property would align the expiration of the cookie
+                // with the expiration of the identity token
+                // UseTokenLifetime = true;
```
Comment (timabell) [inline src/ui/Startup.cs]: I don't understand from the comment why this is commented out but still present

**Comment by timabell** [comment]
It's a bit late but this really should have been two separate branches and PRs

File: src/ui/Startup.cs (hunk 1 above)
Comment (defong) [reply src/ui/Startup.cs]: done

File: src/ui/Startup.cs (hunk 4 above)
Comment (defong) [reply src/ui/Startup.cs]: In our case, we want to align it with the lifetime of the access token, so we have to make a couple of changes.

File: src/ui/Startup.cs (hunk 3 above)
Comment (defong) [reply src/ui/Startup.cs]: Matter of code style... ignoring...

File: src/ui/Startup.cs (hunk 2 above)
Comment (defong) [reply src/ui/Startup.cs]: https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

**Comment by timabell** [comment]
Reject

Failed locally when the provider had no accrediting providers.

**Comment by timabell** [comment]
Fixed it myself....
Going to test with accrediting providers now

File: src/ui/Controllers/OrganisationController.cs [hunk 5]
```diff
@@ -40,39 +42,62 @@ public async Task<IActionResult> Courses(string ucasCode)
...
 
-            return result;
+            return View(model);
+        }
+
+        [HttpPost]
+        [Route("{ucasCode}/about")]
+        public async Task<ActionResult> AboutPost(string ucasCode, OrganisationViewModel model)
+        {
+            var enrichmentModel = (await _manageApi.GetEnrichmentOrganisation(ucasCode)).EnrichmentModel;
```
Comment (timabell) [inline src/ui/Controllers/OrganisationController.cs]: This GET to the API seems like an unnecessary extra round-trip. I think we can just build the model from scratch and POST it.

**Comment by timabell** [comment]
This isn't working when looking at the data received by the api, don't merge it till we've figured this out.

**Comment by timabell** [comment]
The api problem has been resolved and was on the api side.

File: ManageCoursesUi.Tests/Controllers/OrganisationControllerTests.cs [hunk 6]
```diff
@@ -163,7 +173,121 @@ public async Task RequestAccessPost()
...
+
+            var result = await controller.About(ucasCode);
+
+            var viewResult = result as ViewResult;
+
+            Assert.IsNotNull(viewResult);
+            var organisationViewModel = viewResult.ViewData.Model as OrganisationViewModel;
+
+            var tabViewModel = organisationViewModel.TabViewModel;
+            Assert.AreEqual(currentTab, tabViewModel.CurrentTab);
```
Comment (timabell) [inline ManageCoursesUi.Tests/Controllers/OrganisationControllerTests.cs]: Fluent assertions are nicer.

Minor.

File: src/ui/Controllers/OrganisationController.cs [hunk 7]
```diff
@@ -40,39 +42,62 @@ public async Task<IActionResult> Courses(string ucasCode)
...
-            {
-                var providerCourses = institutionCourses.Courses.Where(x => x.AccreditingProviderName == providerName).ToList();
-                var providerId = providerCourses.Select(x => x.AccreditingProviderId).Distinct().SingleOrDefault();//should all be the same
+            var ucasData = (await _manageApi.GetCoursesByOrganisation(ucasCode));
+            var enrichmentModel = (await _manageApi.GetEnrichmentOrganisation(ucasCode)).EnrichmentModel;
 
-                var provider = new Provider
+            var aboutAccreditingTrainingProviders = ucasData.Courses
+                .Where(x =>
+                    false == string.Equals(x.AccreditingProviderId, ucasCode, StringComparison.InvariantCultureIgnoreCase) &&
```
Comment (timabell) [inline src/ui/Controllers/OrganisationController.cs]: use ! to invert

minor

File: src/ui/Controllers/OrganisationController.cs [hunk 8]
```diff
@@ -99,15 +124,57 @@ public async Task<ActionResult> RequestAccessPost(string ucasCode, RequestAccess
...
+            return new RedirectToActionResult("RequestAccess", "Organisation", new { ucasCode });
+        }
+
+        private static Func<AccreditingProviderEnrichment, bool> TrainingProviderMatchesProviderCourse(Course x)
+        {
+            return y => String.Equals(x.AccreditingProviderId,
+            y.UcasInstitutionCode, StringComparison.InvariantCultureIgnoreCase);
+        }
+
+                private List<Provider> GetProviders(InstitutionCourses institutionCourses)
```
Comment (timabell) [inline src/ui/Controllers/OrganisationController.cs]: indentation is off

minor

File: src/ui/Controllers/OrganisationController.cs [hunk 9]
```diff
@@ -99,15 +124,57 @@ public async Task<ActionResult> RequestAccessPost(string ucasCode, RequestAccess
...
+                    Reason = model.Reason,
             });
 
             this.TempData.Add("RequestAccess_To_Name", $"{model.FirstName} {model.LastName}");
 
-            return new RedirectToActionResult("RequestAccess","Organisation", new {ucasCode });
+            return new RedirectToActionResult("RequestAccess", "Organisation", new { ucasCode });
+        }
+
+        private static Func<AccreditingProviderEnrichment, bool> TrainingProviderMatchesProviderCourse(Course x)
```
Comment (timabell) [inline src/ui/Controllers/OrganisationController.cs]: use meaningful param names

minor

File: src/ui/Startup.cs (hunk 2 above)
Comment (timabell) [reply src/ui/Startup.cs]: I'm unconvinced.

Minor

File: src/ui/Startup.cs (hunk 3 above)
Comment (timabell) [reply src/ui/Startup.cs]: indeed. minor.

File: src/ui/Startup.cs (hunk 4 above)
Comment (timabell) [reply src/ui/Startup.cs]: comments in the PR will be lost to the mists of time. I suggest the comment in the code be made clearer.

minor.

**Pull request merged**